<a href="https://colab.research.google.com/github/omkar-dhavale/Titanic-Survival-Prediction/blob/main/Titanic_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# EDA-Exploratory Data Analysis


Importing all the required libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns



---



Loding and Reading the Dataset

In [ ]:
data=pd.read_csv("tested.csv")



---



Understanding the data

In [ ]:
data.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [ ]:
data.shape

(418, 12)

In [ ]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   418.000000  418.000000  418.000000  332.000000  418.000000   
mean   1100.500000    0.363636    2.265550   30.272590    0.447368   
std     120.810458    0.481622    0.841838   14.181209    0.896760   
min     892.000000    0.000000    1.000000    0.170000    0.000000   
25%     996.250000    0.000000    1.000000   21.000000    0.000000   
50%    1100.500000    0.000000    3.000000   27.000000    0.000000   
75%    1204.750000    1.000000    3.000000   39.000000    1.000000   
max    1309.000000    1.000000    3.000000   76.000000    8.000000   

            Parch        Fare  
count  418.000000  417.000000  
mean     0.392344   35.627188  
std      0.981429   55.907576  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   31.500000  
max      9.000000  512.329200

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    object 
 4   Sex          418 non-null    object 
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    object 
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     object 
 11  Embarked     418 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


Conclusion:

Dataset contains 418 records and 12 columns.

Key features like PassengerId, Survived, Pclass, Sex, SibSp, Parch, and Embarked are fully populated.

The remaining columns (Name, Sex, Ticket, Cabin, Embarked) are objects, requiring conversion during preprocessing.

Combining SibSp and Parch to create a FamilySize feature could be useful for further analysis.

---



Checking Whether there are any missing values in dataset.

In [ ]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Conclusion:

Key features like PassengerId, Survived, Pclass, Sex, SibSp, Parch, and Embarked are fully populated.

Approximately 20% of the Age values are missing, so imputation (e.g., using median) is needed.

The Cabin column has significant missing data (only 91 non-null entries), making it less reliable.


---

# 2. Data Cleaning and Feature Engineering

In [ ]:
#We are imputing the null values from its median and mode
data['Age'].fillna(data['Age'].median(), inplace=True)
data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)
data['Fare'].fillna(data['Fare'].median(), inplace=True)

<ipython-input-114-3939b2f2dbb0>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].median(), inplace=True)
<ipython-input-114-3939b2f2dbb0>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', 

In [ ]:
data["Age"].isnull().sum()
data["Embarked"].isnull().sum()
data["Fare"].isnull().sum()

np.int64(0)

In [ ]:
#Cabin column has so many null values, hence we are dropping it
data.drop(columns=['Cabin'], inplace=True)

In [ ]:
data.drop(columns=['Ticket'], inplace=True)
data.drop(columns=['Name'], inplace=True)
data.drop(columns=['PassengerId'], inplace=True)

In [ ]:
#SibSp (Sibling/Spouse Count) and Parch (Parent/Children Count) indicate whether a passenger was traveling alone or with family.
data['FamilySize'] = data['SibSp'] + data['Parch'] + 1

In [ ]:
data.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked', 'FamilySize'],
      dtype='object')

**Encoding:**

In [ ]:
data['Sex'] = data['Sex'].map({'male': 0, 'female': 1})

In [ ]:
data['Sex'].head()

0    0
1    1
2    0
3    0
4    1
Name: Sex, dtype: int64

In [ ]:
data=pd.get_dummies(data,columns=['Embarked'], drop_first=True)

# 3. Feature Selection and Splitting the Data


---



Selecting the Target variable and Independent Variable

In [ ]:
X = data.drop('Survived', axis=1)
y = data['Survived']

Splitting the Dataset

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
columns_to_scale = ['Age', 'Fare', 'FamilySize']
preprocessor = ColumnTransformer(transformers=[
    ('num', StandardScaler(), columns_to_scale)
], remainder='passthrough')

# Model Building and Evaluation





**Logistic Regression**

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
lr_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(max_iter=1000, random_state=42))
])

Training the model

In [ ]:
lr_pipeline.fit(X_train, y_train)
y_pred_lr = lr_pipeline.predict(X_test)

In [ ]:
y_pred_lr=pd.DataFrame(y_pred_lr)
y_pred_lr.head()

0
0  0
1  1
2  0
3  0
4  1

**Evaluating the model**

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
print("Accuracy:", accuracy_score(y_test, y_pred_lr))
print("Classification Report:\n", classification_report(y_test, y_pred_lr))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_lr))

Accuracy: 1.0
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       1.00      1.00      1.00        34

    accuracy                           1.00        84
   macro avg       1.00      1.00      1.00        84
weighted avg       1.00      1.00      1.00        84

Confusion Matrix:
 [[50  0]
 [ 0 34]]




---



---

